agent/build.gradle.kts

```diff
@@ -13,98 +13,81 @@ java {
   withSourcesJar()
 }
 
-// dependencies that already are relocated and will be moved to inst/ (agent classloader isolation)
-val isolateLibs by configurations.creating
-// dependencies that will be relocated
-val relocateLibs by configurations.creating
-// dependencies that will be included as they are
-val includeAsIs by configurations.creating
-
-configurations {
-  named("compileOnly") {
-    extendsFrom(includeAsIs)
-  }
+// this configuration collects libs that will be placed in the bootstrap classloader
+val bootstrapLibs by configurations.creating {
+  isCanBeResolved = true
+  isCanBeConsumed = false
+}
+// this configuration collects libs that will be placed in the agent classloader, isolated from the instrumented application code
+val javaagentLibs by configurations.creating {
+  isCanBeResolved = true
+  isCanBeConsumed = false
+}
+// this configuration stores the upstream agent dep that's extended by this project
+val upstreamAgent by configurations.creating {
+  isCanBeResolved = true
+  isCanBeConsumed = false
 }
 
 val otelInstrumentationVersion: String by extra
 
-// TODO: remove those three lines after shadowJar plugin fixes its afterEvaluate() calls
-// since gradle 7.0 calling afterEvaluate() on already configured project throws an error, and this is exactly
-// what the shadowJar plugin seems to do - calling evaluationDependsOn() for all shadowed projects seems to delay
-// the configuration until after the plugin is applied
-evaluationDependsOn(":custom")
-evaluationDependsOn(":instrumentation")
-evaluationDependsOn(":profiler")
-
 dependencies {
-  isolateLibs(project(":custom", configuration = "shadow"))
-  isolateLibs(project(":instrumentation", configuration = "shadow"))
-  isolateLibs(project(":profiler", configuration = "shadow"))
+  bootstrapLibs(project(":bootstrap"))
 
-  relocateLibs(project(":bootstrap"))
+  javaagentLibs(project(":custom"))
+  javaagentLibs(project(":instrumentation"))
+  javaagentLibs(project(":profiler"))
 
-  includeAsIs("io.opentelemetry.javaagent:opentelemetry-javaagent:${otelInstrumentationVersion}:all")
-}
-
-fun isolateAgentClasses (jars: Iterable<File>): CopySpec {
-  return copySpec {
-    jars.forEach {
-      from(zipTree(it)) {
-        into("inst")
-        rename("(^.*)\\.class\$", "\$1.classdata")
-      }
-    }
-  }
+  upstreamAgent("io.opentelemetry.javaagent:opentelemetry-javaagent:${otelInstrumentationVersion}:all")
 }
 
 tasks {
-  compileJava {
-    options.release.set(8)
-  }
-
   jar {
     enabled = false
   }
 
-  // includes everything except otel agent
-  val relocateAndIsolate by registering(ShadowJar::class) {
-    dependsOn(":custom:shadowJar")
-    dependsOn(":instrumentation:shadowJar")
-    dependsOn(":profiler:shadowJar")
-    dependsOn(":bootstrap:jar")
-    dependsOn(":agent:jar")
+  val relocateJavaagentLibs by registering(ShadowJar::class) {
```

Comment:
I think we need some explanatory documentation for all this. I would mention exact steps that are happening: first, this task shades, then next Copy task isolates, then last shadow task adds upstream agent and bootstrap. And also that actual shading rules come from the convention. This last piece got me confused for a couple of minutes...

Reply:
Good point, done.